backends/tfhe-cuda-backend/cuda/include/zk/zk_utilities.h

```diff
@@ -0,0 +1,217 @@
+#ifndef ZK_UTILITIES_H
+#define ZK_UTILITIES_H
+
+#include "../integer/integer_utilities.h"
+#include "integer/integer.cuh"
+#include <cstdint>
+
+template <typename Torus> struct zk_expand {
+  int_radix_params params;
+  uint32_t num_lwes;
+  uint32_t num_compact_lists;
+
+  int_radix_lut<Torus> *message_and_carry_extract_luts;
+
+  Torus *tmp_expanded_lwes;
+  Torus *tmp_ksed_small_to_big_expanded_lwes;
+
+  uint32_t *h_compact_list_length_per_thread;
+  uint32_t *d_compact_list_length_per_thread;
+
+  uint32_t *d_lwe_compact_input_indexes;
+
+  zk_expand(cudaStream_t const *streams, uint32_t const *gpu_indexes,
+            uint32_t gpu_count, int_radix_params params,
+            const uint32_t *num_samples_per_compact_list,
+            const bool *is_boolean_array, uint32_t num_compact_lists,
+            bool allocate_gpu_memory)
+      : params(params), num_compact_lists(num_compact_lists) {
+
+    num_lwes = 0;
+    for (int i = 0; i < num_compact_lists; i++) {
+      num_lwes += num_samples_per_compact_list[i];
+    }
+
+    if (params.carry_modulus != params.message_modulus) {
+      PANIC("GPU backend requires carry_modulus equal to message_modulus")
+    }
+
+    if (allocate_gpu_memory) {
+
+      h_compact_list_length_per_thread =
+          (uint32_t *)malloc(num_lwes * sizeof(uint32_t));
+      d_compact_list_length_per_thread = (uint32_t *)cuda_malloc_async(
+          num_lwes * sizeof(uint32_t), streams[0], gpu_indexes[0]);
+
+      // This loop iterates over the total number of LWEs (num_lwes) and assigns
+      // the compact list size per thread to each LWE based on the provided
```

Comment:
What do you mean here by "per thread"? Isn't it the length per list that each LWE belongs to rather?

Reply:
You are right, this is not "per thread".